### PR TITLE
Allow shard setting with comments

### PR DESCRIPTION
**Pull request by jmeagher**
# What

Allows shard selection by the client to come in via comments like `/* shard_id: 1 */ select * from foo;`

# Why

We're using a setup in Ruby that makes it tough or impossible to inject commands on the connection to set the shard before it gets to the "real" SQL being run. Instead we have an updated PG adapter that allows injection of comments before each executed SQL statement. We need this support in pgcat in order to keep some complex shard picking logic in Ruby code while using pgcat for connection management. 

# Local Testing

Run postgres and pgcat with the default options. Run `psql < tests/sharding/query_routing_setup.sql` to setup the database for the tests and run `./tests/pgbench/external_shard_test.sh` as often as needed to exercise the shard setting comment test.

This was developed with @drdrsh

### CODE REVIEW

**Comment by levkk** [comment]
Hey there John, nice to be working together again!

I think you're on the right track with this PR. We've been thinking about this functionality for a while now and this is definitely a welcome change.

A few pointers to make this work:

##### `QueryRouter`
Take a look at `QueryRouter`, specifically `try_execute_command`. It could be nice to incorporate this there. You'll find a lot of similar regexes living there as well.

##### Test suite
The existing test suite is pretty comprehensive, and you'll also note the presence of Ruby and Python suits (Mostafa likes Ruby while Zain likes Python, personally I'm more of a bash guy, but everyone is welcome here); pick your favorite.

`query_routing_setup.sql` does all of the setup so you just need to write your own SQL tests if desired. You'll note the presence of partitions that should validate that your query routing addition works as intended (otherwise there will be insertion errors due to partition bounds). They should have issues working locally, not sure why Circle is broken (isn't it always in every project ever...). 

##### Backwards compatibility
PgCat is being used in production by a few organizations now, so any functionality that changes core behavior should be optional; the easiest way to make that work is to make it config-driven.

In your case, I'd suggest a config to tell the query parser to either use the existing SQL AST parser or your Regex approach, maybe something like `query_parser_mode = "regex"` or `query_parser_mode = "ast_parser"`. Then the regex should also be configurable, as you may suspect different people write their SQL comments differently. A bit of historical context: https://github.com/levkk/pgcat/issues/168

Let me know if I can help further. Welcome to the project!

**Comment by jmeagher** [comment]
I did some cleanup on this to move the logic somewhere it made more sense. I also made it configurable and optional by default. It now also support setting a sharding_key to help support https://github.com/levkk/pgcat/issues/168. If other special commands are needed via comments it should be easy to expand this to include other options like role. 

Let me know how this looks.

File: src/pool.rs
Comment (levkk) [inline src/pool.rs:301]: It would be good to validate this in `Config#validate` function so the user gets a nice error on startup instead of a panic.

File: src/query_router.rs
Comment (levkk) [inline src/query_router.rs:137]: I think you want to early return `None` here so the rest of the command parsing logic doesn't run for no reason.

File: src/query_router.rs
Comment (levkk) [inline src/query_router.rs:151]: Same here.

File: src/query_router.rs
Comment (levkk) [inline src/query_router.rs:156]: Would be nice to early return None here as well.

File: src/query_router.rs
Comment (levkk) [inline src/query_router.rs:160]: You can delete this part now, you're already checking for valid `P` or `Q` messages above.

File: src/config.rs
Comment (levkk) [inline src/config.rs:387]: It would be nice to add these to pgcat.toml which we use for illustration and documentation purposes.

File: src/query_router.rs
Comment (jmeagher) [reply src/query_router.rs:137]: Ok, I think I misunderstood what this code is doing. The whole rest of the function is only about the special command parsing. I guess it's assumed that if you're passing the shard information in a comment then it won't need to look for any special commands? I'll put in the `return None` for these when it successfully finds sharding information in the new code.

File: src/query_router.rs
Comment (jmeagher) [reply src/query_router.rs]: I moved the 'Q' check to here. I think it should continue processing below for 'P' code if no sharding information is found.

File: pgcat.toml
Comment (jmeagher) [inline pgcat.toml:90]: These are commented out because when they were uncommented it caused the Ruby TOML parser to fail when loading the file. These aren't used in tests right now, so it should be fine.

File: src/query_router.rs
Comment (levkk) [inline src/query_router.rs:162]: You won't hit this check because the `match` branch accepts both `P` and `Q`.

File: pgcat.toml
Comment (levkk) [reply pgcat.toml:90]: We will need to figure this out because otherwise this feature won't be tested in end-to-end tests we have in Ruby. I think it's fine to test it with Bash to start, but the Ruby test suit  protects us against bad things and we want it to run there too.

File: src/query_router.rs
Comment (jmeagher) [reply src/query_router.rs:162]: oh yeah, so I think that means the old 'Q' check below this is still needed

File: src/query_router.rs
Comment (jmeagher) [reply src/query_router.rs:160]: I moved the original code back. I don't think this should return None here since there could be a mixed-case use of having comment-based-sharding enabled on a pool, but still wanting to send over the command based shard setting commands.

Comment (levkk) [reply src/query_router.rs:160]: Works!

**Review by levkk — APPROVED** [review]
The LOC LB test is flakey, rerunning it usually helps. We have an action item to fix this. The PR looks good to me to merge. Eventually we would want some tests in bash/Ruby/Python so we don't get hit with a regression. I'll let @drdrsh take a look before merging.

File: src/query_router.rs
Comment (drdrsh) [inline src/query_router.rs:120]: I think we can avoid this entire branch when the number of shards is 1, we can get that from `self.pool_settings.shards`. This will make this check even cheaper for the base, unsharded use case

Comment (drdrsh) [reply src/query_router.rs:120]: Scratch that, the lack of the regex in the confs should be enough of a gate

**Comment by drdrsh** [comment]
We'll test this change internally. I will merge it once we are done @levkk

File: src/config.rs
Comment (jmeagher) [inline src/config.rs:327]: Having the new fields below user and shard broke our internal tool that generates the TOML config. Since this is a pretty odd error that we hit I'm leaving behind the breadcrumb that got us to fix it.

**Pull request merged**
